dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/token/TokenResponseValidator.java

```diff
@@ -190,14 +219,41 @@ private JSONObject getProviderDiscoveryMetadadata() {
 
     /**
      * @return
+     * @throws OidcDiscoveryException
      */
-    private String getJwkUri() {
-        String jwkuri = null;
-        jwkuri = clientConfig.getProviderMetadata().getJwksURI();
-        if ((jwkuri == null || jwkuri.isEmpty()) && getProviderDiscoveryMetadadata() != null) {
-            jwkuri = (String) (this.discoveredProviderMetadata.get("jwks_uri"));
+    private String getJwksUri() throws OidcDiscoveryException {
+        String jwksUri = getJwksUriFromProviderMetadata();
+        if (jwksUri != null && !jwksUri.isEmpty()) {
+            return jwksUri;
+        }
+        return getJwksUriFromDiscoveryMetadata();
+    }
+
+    private String getJwksUriFromProviderMetadata() {
+        OidcProviderMetadata providerMetadata = clientConfig.getProviderMetadata();
+        if (providerMetadata != null) {
+            String jwksUri = providerMetadata.getJwksURI();
+            if (jwksUri != null && !jwksUri.isEmpty()) {
+                if (tc.isDebugEnabled()) {
+                    Tr.debug(tc, "JWK Set Uri found in the provider metadata: [" + jwksUri + "]");
+                }
+                return jwksUri;
+            }
+        }
+        return null;
+    }
+
+    private String getJwksUriFromDiscoveryMetadata() throws OidcDiscoveryException {
+        String jwksUri = null;
+        JSONObject providerDiscoveryMetadata = getProviderDiscoveryMetadadata();
+        if (providerDiscoveryMetadata != null) {
+            jwksUri = (String) providerDiscoveryMetadata.get(OidcDiscoveryConstants.METADATA_KEY_JWKS_URI);
+        }
+        if (jwksUri == null || jwksUri.isEmpty()) {
+            String nlsMessage = Tr.formatMessage(tc, "DISCOVERY_METADATA_MISSING_VALUE", OidcDiscoveryConstants.METADATA_KEY_JWKS_URI);
+            throw new OidcDiscoveryException(clientConfig.getClientId(), clientConfig.getProviderURI(), nlsMessage);
```

Comment:
If algorithm is HS, then we don't need jwksUri to do the signature validation

Reply:
I agree - I don't think a missing JWKS URI is necessarily a failure.

Reply:
the jwks uri is required in the jakarta spec and in the oidc discovery spec (does not specify an exception for hs* algorithms), so an error should be thrown here.

https://github.com/jakartaee/security/blob/master/api/src/main/java/jakarta/security/enterprise/authentication/mechanism/http/openid/OpenIdProviderMetadata.java#L74-L84

https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata